### ISSUE TITLE
BUG:Dataframe assignment failure

### ISSUE THREAD

**Issue by quanbingDG** [pandas-dev/pandas#39088]
- [ ] I have checked that this issue has not already been reported.

- [ ] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
# Your code here
import pandas as pd
data = pd.DataFrame(np.array(range(12)).reshape(3,4))
data['1_xxx'] = 2
```

#### Problem description

When a column name consists of a number and a string. that triggers the bug.

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit           : d9fff2792bf16178d4e450fe7384244e50635733
python           : 3.7.9.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 20.2.0
Version          : Darwin Kernel Version 20.2.0: Wed Dec  2 20:39:59 PST 2020; root:xnu-7195.60.75~1/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : zh_CN.UTF-8
LOCALE           : zh_CN.UTF-8

pandas           : 1.2.0
numpy            : 1.19.2
pytz             : 2020.5
dateutil         : 2.8.1
pip              : 20.3.3
setuptools       : 51.0.0.post20201207
Cython           : 0.29.21
pytest           : 6.2.1
hypothesis       : None
sphinx           : 3.4.2
blosc            : None
feather          : None
xlsxwriter       : 1.3.7
lxml.etree       : 4.6.2
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.19.0
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : 1.3.2
fsspec           : 0.8.3
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.2
numexpr          : 2.7.2
odfpy            : None
openpyxl         : 3.0.5
pandas_gbq       : None
pyarrow          : None
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : 1.5.2
sqlalchemy       : 1.3.21
tables           : 3.6.1
tabulate         : None
xarray           : None
xlrd             : 2.0.1
xlwt             : 1.3.0
numba            : 0.51.2
[paste the output of ``pd.show_versions()`` here leaving a blank line after the details tag]

</details>

**Comment by quanbingDG** [comment]
it works on version 1.1.0

**Comment by MarcoGorelli** [comment]
thanks for the report @quanbingDG - this is fixed on `master`

**Comment by MarcoGorelli** [comment]
```
6e707f2f68a4ed7fd990438bc489374641374cff is the first bad commit
commit 6e707f2f68a4ed7fd990438bc489374641374cff
Author: patrick <61934744+phofl@users.noreply.github.com>
Date:   Fri Jan 8 01:19:43 2021 +0100

    Fix regression in setitem when expanding DataFrame with specific column name format (#39023)

 doc/source/whatsnew/v1.2.1.rst              | 1 +
 pandas/core/dtypes/common.py                | 2 +-
 pandas/tests/dtypes/test_common.py          | 5 +++++
 pandas/tests/frame/indexing/test_setitem.py | 7 +++++++
 4 files changed, 14 insertions(+), 1 deletion(-)
bisect run success
```

looks like this was fixed intentionally in #39023, and there's already a test, so I think we're good to close
